src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -127,27 +148,31 @@ HeapWord* ShenandoahFreeSet::allocate_single(ShenandoahAllocRequest& req, bool&
       if (result != NULL) {
         return result;
       }
-      // Then try a free region that is dedicated to GC allocations.
-      result = allocate_with_affiliation(FREE, req, in_new_region);
-      if (result != NULL) {
-        return result;
+      if (allow_new_region) {
+        // Then try a free region that is dedicated to GC allocations.
+        result = allocate_with_affiliation(FREE, req, in_new_region);
+        if (result != NULL) {
+          return result;
+        }
       }
 
       // No dice. Can we borrow space from mutator view?
       if (!ShenandoahEvacReserveOverflow) {
         return NULL;
       }
 
-      // Try to steal an empty region from the mutator view.
-      for (size_t c = _mutator_rightmost + 1; c > _mutator_leftmost; c--) {
-        size_t idx = c - 1;
-        if (is_mutator_free(idx)) {
-          ShenandoahHeapRegion* r = _heap->get_region(idx);
-          if (can_allocate_from(r)) {
-            flip_to_gc(r);
-            HeapWord *result = try_allocate_in(r, req, in_new_region);
-            if (result != NULL) {
-              return result;
+      if (allow_new_region) {
+        // Try to steal an empty region from the mutator view.
+        for (size_t c = _mutator_rightmost + 1; c > _mutator_leftmost; c--) {
+          size_t idx = c - 1;
+          if (is_mutator_free(idx)) {
+            ShenandoahHeapRegion* r = _heap->get_region(idx);
+            if (can_allocate_from(r)) {
+              flip_to_gc(r);
```

Comment:
There's no flip to mutator because we do not allow a mutator allocation request to take memory that had been set aside for use by the collector (for evacuations).  If a mutator alloc "fails", we can stall that single mutating thread.  If a GC evacuation fails, we have to force all threads into a safepoint so that we can perform a FULL GC.  This is the reason we don't allow mutators to flip_to_mutator().